aptos-move/framework/aptos-stdlib/sources/pool_u64.move

```diff
@@ -0,0 +1,188 @@
+    public fun balance(pool: &Pool, shareholder: address): u64 {
+        let num_shares = shares(pool, shareholder);
+        shares_to_amount(pool, num_shares)
+    }
+
+    /// Return the list of shareholders in `pool`.
+    public fun shareholders(pool: &Pool): vector<address> {
+        pool.shareholders
+    }
+
+    /// Return the number of shareholders in `pool`.
+    public fun shareholders_count(pool: &Pool): u64 {
+        vector::length(&pool.shareholders)
+    }
+
+    /// Update `pool`'s total balance of coins.
+    public fun update_total_coins(pool: &mut Pool, new_total_coins: u64) {
+        pool.total_coins = (new_total_coins as u128);
+    }
+
+    /// Allow an existing or new shareholder to add their coins to the pool in exchange for new shares.
+    public fun buy_in(pool: &mut Pool, shareholder: address, coins_amount: u64): u64 {
+        let num_shares = amount_to_shares(pool, coins_amount);
+        pool.total_coins = pool.total_coins + (coins_amount as u128);
+        pool.total_shares = pool.total_shares + (num_shares as u128);
+        add_shares(pool, shareholder, num_shares);
+
+        num_shares
+    }
+
+    /// Add the number of shares directly for `shareholder` in `pool`.
+    /// This would dilute other shareholders if the pool's balance of coins didn't change.
+    public fun add_shares(pool: &mut Pool, shareholder: address, num_shares: u64): u64 {
+        let num_shares_u128 = (num_shares as u128);
+        if (contains(pool, shareholder)) {
+            let existing_shares = simple_map::borrow_mut(&mut pool.shares, &shareholder);
+            *existing_shares = *existing_shares + num_shares_u128;
+
+            (*existing_shares as u64)
+        } else {
+            assert!(
+                vector::length(&pool.shareholders) < pool.shareholders_limit,
+                error::invalid_state(ETOO_MANY_SHAREHOLDERS),
+            );
+
+            vector::push_back(&mut pool.shareholders, shareholder);
+            simple_map::add(&mut pool.shares, shareholder, num_shares_u128);
+
+            num_shares
+        }
+    }
+
+    /// Allow `shareholder` to redeem their shares in `pool` for coins.
+    public fun redeem_shares(pool: &mut Pool, shareholder: address, num_shares: u64): u64 {
+        assert!(contains(pool, shareholder), error::invalid_argument(ESHAREHOLDER_NOT_FOUND));
+        assert!(shares(pool, shareholder) >= num_shares, error::invalid_argument(EINSUFFICIENT_SHARES));
+
+        let num_coins = shares_to_amount(pool, num_shares);
+        pool.total_coins = pool.total_coins - (num_coins as u128);
+        pool.total_shares = pool.total_shares - (num_shares as u128);
+        deduct_shares(pool, shareholder, num_shares);
+
+        num_coins
+    }
+
+    /// Directly deduct `shareholder`'s number of shares in `pool` and return the number of remaining shares.
+    public fun deduct_shares(pool: &mut Pool, shareholder: address, num_shares: u64): u64 {
```

Comment:
not public

Reply:
Same as above. This allows the owner of the pool to update shares for a specific shareholder which can be useful if this is an update outside of redemption.

Reply:
so you remove shares but leave coins in, so it feels very awkward to me...

Reply:
Made private in the other PR